## Project 5
- created 5-16-22 by GTP
- https://docs.google.com/document/d/1LIJTlCsx54zIG5sOX3heSj00YqdzhZ9c/edit
- *Description*: BSEED has entered a lot of data into free text fields within Accela. Would be useful to find ways to scrape and organize this data so it is useable. Unit data and Certificates of Occupancy are some of our biggest gaps. This might be a way to use administrative data to version and validate 2020 data.
- Technical Skill Level: Medium-High. Skilled at applying Regex to text strings using SQL and/or Python. Experience working with geospatial data, in ArcGIS or otherwise.
- Scope: There are 595 records in the Certificates of Occupancy dataset and 5,930 records in the Certificates of Compliance dataset. Depending on skill level, this could take 6-8 weeks.
- Inputs: Certificates of Compliance, Certificates of Occupancy, Rental Registration data
- General Process:
- Use GIS or Base Units Explorer tool to link Certificates of Occupancy to specific building ids, to create timestamps for when a building was ready for occupants.
- Geocode the addresses in the Certificate of Compliance and Rental Registration datasets and note any addresses that can’t be matched through a manual rematching process and may be missing altogether from the database.


In [1]:
#import data libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import math
import numbers
import decimal
#import data science packages
import scipy
import scipy.stats as stats

np.random.seed(222)
%matplotlib inline

In [13]:
#import geographic analysis libraries
import geopandas as gpd
from geopandas import GeoDataFrame
import shapely as shp
from shapely.geometry import Point
from shapely.geometry import shape
import os
import re
from fiona.crs import from_epsg
import pysal as ps
import re
from googlemaps import Client as GoogleMaps
import googlemaps
import gmaps

In [15]:
# This is where we will need the API key
gmaps = googlemaps.Client(key=os.environ['GOOGLE_GEOCODER_API'])

In [6]:
#set crs for entire analysis
crs = {'init': 'epsg:4326'}

### data sources

Certificates of Occupancy: https://data.detroitmi.gov/datasets/certificates-of-occupancy-1/explore
- BSEED says that new building (or a rehabbed / renovated building) has satisfied their requirements for habitation, and people can move in / it is ready for occupancy
- note: alice says that this can be issued for individual floors 
- _goal_: deliverable should be a table that is the certificate of occupancy number (record_id) and building footprint ids - sometimes these are 1 to 1 and then sometimes multiple occupancy numbers might relate to a single id
- the census challenge is interested in having this as a record to when exactly a new building was technically 'habitable' - the "birthdate" of the property in terms of occupancy

Certificates of Compliance: https://data.detroitmi.gov/datasets/certificates-of-compliance-1/explore
- this is for properties to be certified as 'compliant' by the city
- Alice has access to the dataset of compliance that has 'description' - which should contain additional details..?
- I think this will be more trying to geocode the ones that don't have lat/lon
- _goal_: there's 33 that didn't geocode - goal is to geocode these and then give description if couldn't geocode

Rental Registrations: https://data.detroitmi.gov/datasets/rental-statuses-1/explore
- (6-1-22): I'll address this next week with Alice on our next call

Base Units: https://base-units-detroitmi.hub.arcgis.com/datasets/detroitmi::units-1/about
- jimmy mcbroom put this together

https://cityofdetroit.github.io/base-unit-tools/explorer

In [7]:
compliance_gdf = gpd.read_file('../data/Certificates_Of_Compliance/Certificates_Of_Compliance.shp')

In [8]:
len(compliance_gdf)

6071

In [9]:
len(compliance_gdf[compliance_gdf['geometry'].isna()])

33

In [10]:
compliance_gdf_nogeocode = compliance_gdf[compliance_gdf['geometry'].isna()]

In [30]:
compliance_gdf_nogeocode['street_dir'] = compliance_gdf_nogeocode['street_dir'].apply(lambda x: str(x))
compliance_gdf_nogeocode['street_dir'] = pd.Series(compliance_gdf_nogeocode['street_dir']).str.replace('None', '', regex=False)
compliance_gdf_nogeocode['street_typ'] = compliance_gdf_nogeocode['street_typ'].apply(lambda x: str(x))
compliance_gdf_nogeocode['street_typ'] = pd.Series(compliance_gdf_nogeocode['street_typ']).str.replace('None', '', regex=False)

/Users/geoffperrin/miniconda3/lib/python3.8/site-packages/geopandas/geodataframe.py:1322: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  super(GeoDataFrame, self).__setitem__(key, value)


In [31]:
compliance_gdf_nogeocode['address_for_geocode'] = compliance_gdf_nogeocode['street_num'].astype(str) + ' ' +\
                                                  compliance_gdf_nogeocode['street_dir'].astype(str) + ' ' +\
                                                  compliance_gdf_nogeocode['street_nam'].astype(str) + ' ' +\
                                                  compliance_gdf_nogeocode['street_typ'].astype(str) + ' ' +\
                                                  'DETROIT MI'

In [36]:
def google_geocode(address_to_geocode):
    geocode_result = gmaps.geocode(address_to_geocode)
    lat = geocode_result[0]['geometry']['location']['lat']
    lon = geocode_result[0]['geometry']['location']['lon']
    return (lat, lon)

In [ ]:
def return_lat(lat_lon):
    return lat_lon[0]

def return_lon(lat_lon):
    return lat_lon[1]

In [ ]:
compliance_gdf_nogeocode['lat_lon'] = compliance_gdf_nogeocode['address_for_geocode'].apply(compliance_gdf_nogeocode)
compliance_gdf_nogeocode['new_lat'] = compliance_gdf_nogeocode['lat_lon'].apply(return_lat)
compliance_gdf_nogeocode['new_lon'] = compliance_gdf_nogeocode['lat_lon'].apply(return_lon)

### Note: need to set up google geocoder billing account
- link: https://console.cloud.google.com/billing
- because i'm currently in gdl, mx, was denied right now. trying to figure out how to get unblocked, waiting for google's 'security review'

In [37]:
google_geocode('16157 FIELDING DETROIT MI')

ApiError: REQUEST_DENIED (You must enable Billing on the Google Cloud Project at https://console.cloud.google.com/project/_/billing/enable Learn more at https://developers.google.com/maps/gmp-get-started)

### Compliance df goal:
- export list of non-geocoded compliance to google sheets
- try to geocode these manually
- if can't geocode, then add description column as to why
- google sheet: https://docs.google.com/spreadsheets/d/1agyVFNR8gtoQabLZHSyjIecW8lv3LYlD_7BvXUEZRSM/edit#gid=0

In [9]:
compliance_gdf_nogeocode.to_csv('../data/exports/compliance_gdf_nogeocode.csv')

## Occupancy DF
- notes: descriptio column has free text that we could leverage to fill in empty geometry cells
- "AKA 2327 Trumbull Ave" is an example of the free text - as in, AKA "address" is a common pattern, but this already exists / has been extracted into the street_num / street_nam columns - maybe those just have to be geocoded

- we're looking for the relationship between housing units and certificates of occupancy. That relationship is often mediated by a building id (use https://cityofdetroit.github.io/base-unit-tools/explorer?id=3263&type=buildings&streetview=true) / there's a xwalk that alice will send over

- deliverable should be a table that is the certificate of occupancy number (record_id) and building footprint ids - sometimes these are 1 to 1 and then sometimes multiple occupancy numbers might relate to a single id

- the census challenge is interested in having this as a record to when exactly a new building was technically 'habitable' - the "birthdate" of the property in terms of occupancy

In [11]:
occupancy_gdf = gpd.read_file('../data/Certificates_Of_Occupancy/Certificates_Of_Occupancy.shp')

In [12]:
len(occupancy_gdf[occupancy_gdf['geometry'].isna()])/len(occupancy_gdf)

0.16166666666666665

In [13]:
len(occupancy_gdf[occupancy_gdf['geometry'].isna()])

97

In [14]:
occupancy_gdf[occupancy_gdf['geometry'].isna()]

record_id  street_num street_dir          street_nam street_typ  \
4    BLD2017-06254        8401       None         WOODMONT 14       None   
12   BLD2019-03478        1541       None         Fisher Frwy       None   
23   BLD2019-04976        2327       None            Trumbull       None   
25   BLD2018-05987        1230       None             LIBRARY       None   
36   BLD2019-00054        2817       None               BRUSH       None   
..             ...         ...        ...                 ...        ...   
576  BLD2017-09621         286       None              ALFRED       None   
582  BLD2019-00679        2807       None               Brush       None   
592  BLD2020-00477        4501       None           St. Aubin       None   
595  BLD2019-02529         692       None           AMSTERDAM       None   
596  BLD2019-00722         500       None  RENAISSANCE CENTER       None   

                                            descriptio       status  \
4    SEE BLD2017-00831\nAKA 8227, 8237, 8243, 8251 ...  CofO Issued   
12   AKA 1541 W. Fisher Freeway Unit 29. Per BZA #4...  CofO Issued   
23   AKA 2327 Trumbull Ave. Unit 20. Per BZA #4-18,...  CofO Issued   
25           PERMANENT CERTIFICATE OF OCCUPANCY ISSUED  CofO Issued   
36   ERECT A 4 STORY, TOWNHOME AS PER EPLANS AND CE...  CofO Issued   
..                                                 ...          ...   
576  (AKA 292 ALFRED) ERECT A 3 STORY, TOWNHOUSE W/...  CofO Issued   
582  New residential structure 8  units 4 story tow...  CofO Issued   
592  Revision to BLD2019-02421 per plans.\r\n(Const...  CofO Issued   
595  Revision to BLD2018-07772 to reflect changes t...  CofO Issued   
596  Change of Occupancy to expand an existing Day ...  CofO Issued   

     date_statu parcel_id  lon  lat  ObjectId geometry  
4    2019-02-23      None  NaN  NaN         5     None  
12   2020-08-12      None  NaN  NaN        13     None  
23   2020-10-15      None  NaN  NaN        24     None  
25   2019-04-22      None  NaN  NaN        26     None  
36   2021-06-01      None  NaN  NaN        37     None  
..          ...       ...  ...  ...       ...      ...  
576  2021-09-07      None  NaN  NaN       577     None  
582  2021-05-28      None  NaN  NaN       583     None  
592  2021-02-24      None  NaN  NaN       593     None  
595  2020-08-19      None  NaN  NaN       596     None  
596  2019-07-30      None  NaN  NaN       597     None  

[97 rows x 13 columns]

In [15]:
occupancy_gdf['descriptio'][occupancy_gdf['record_id']=='BLD2021-02415'].values[0]

'(AKA 3321 Cochrane) Construct (11) unit Rowhouse building and Accessory Garages per BZA (41-19) & (SLU2019-00020) per Plans.\r\n(Permit reviewed under BLD2019-03775)'

In [16]:
occupancy_gdf[occupancy_gdf['record_id']=='BLD2021-02415']

record_id  street_num street_dir street_nam street_typ  \
467  BLD2021-02415        3303       None   COCHRANE       None   

                                            descriptio       status  \
467  (AKA 3321 Cochrane) Construct (11) unit Rowhou...  CofO Issued   

     date_statu     parcel_id        lon        lat  ObjectId  \
467  2021-10-11  08006537.001 -83.074239  42.339555       468   

                       geometry  
467  POINT (-83.07424 42.33956)

In [17]:
len(occupancy_gdf)

600

In [18]:
occupancy_gdf_empty = occupancy_gdf[occupancy_gdf['geometry'].isna()]

In [20]:
occupancy_gdf_empty[occupancy_gdf_empty['record_id']=='BLD2019-00680']

record_id  street_num street_dir street_nam street_typ  \
84  BLD2019-00680        2809       None      Brush       None   

                                           descriptio       status  \
84  Erect  4 story , 8 unit townhomes as per eplan...  CofO Issued   

    date_statu parcel_id  lon  lat  ObjectId geometry  
84  2021-04-05      None  NaN  NaN        85     None

In [21]:
occupancy_gdf_empty['descriptio'][occupancy_gdf_empty['record_id']=='BLD2019-00680'].values

array(['Erect  4 story , 8 unit townhomes as per eplans w/ a certificate of appropriateness'],
      dtype=object)

In [22]:
occupancy_gdf_empty['descriptio'][occupancy_gdf_empty['record_id']=='BLD2020-01564'].values

array(['Modify previous Change of Use Permit to Provisioning Center by adding grow facility; changes to the restroom facilities.'],
      dtype=object)

In [23]:
occupancy_gdf_empty['descriptio'][occupancy_gdf_empty['record_id']=='BLD2019-00033'].values

array(['INTERIOR ALTERATIONS TO ESTABLISH USE FOR TENANT SPACE AS COSMETIC RETAIL\nPERMANENT CERTIFICATE OF OCCUPANCY ISSUED (03-20-2019)'],
      dtype=object)

In [24]:
occupancy_gdf_empty['descriptio'][occupancy_gdf_empty['record_id']=='BLD2020-04413'].values[0]

'Interior alterations per plans.(1500 E. Woodbridge Suite address per plans, Separate Tenant Build-Out Permit required to establish Occupancy). Subject to all Applicable Federal, State, and Local Executive Orders.\r\n(AKA 1583 Franklin)'

- note: this building is at the corner of e. woodbridge and franklin (hence the aka 1583 franklin)

In [25]:
occupancy_gdf_empty[occupancy_gdf_empty['record_id']=='BLD2020-04413']

record_id  street_num street_dir  street_nam street_typ  \
393  BLD2020-04413        1522       None  WOODBRIDGE       None   

                                            descriptio       status  \
393  Interior alterations per plans.(1500 E. Woodbr...  CofO Issued   

     date_statu parcel_id  lon  lat  ObjectId geometry  
393  2021-04-12      None  NaN  NaN       394     None

In [26]:
occupancy_gdf_empty['descriptio'][occupancy_gdf_empty['record_id']=='BLD2019-04976'].values

array(["AKA 2327 Trumbull Ave. Unit 20. Per BZA #4-18, Construct 34' L X 21' W X 37' H Townhouse per plans."],
      dtype=object)

In [27]:
occupancy_gdf_empty['descriptio'][occupancy_gdf_empty['record_id']=='BLD2017-06240'].values

array(['AKA 8032, 8040, 8046, 8056 MEMORIAL. ERECTION OF ONE 4 UNIT ONE STORY WOOD FRAMED TOWNHOUSE AS PER PLANS. SEE BLD2017-00831 FOR MASTER SET OF PLANS.'],
      dtype=object)

## Rental Statuses / Rental Registrations
### Notes:
- https://data.detroitmi.gov/datasets/rental-statuses-1/explore
- no descriptions / free text, so could be a limitation on using that to geocode
- BSEED data overly relies on parcels, so they can't really take into account buildings that have both rentals _and_ owner occupied units, like 120 seward
- so, those mixed buildings will not all be in this data, and the counts will be low. 120 seward only has 2 observations
- bseed probably only cares about building level data i.e. is the building sound and safe
- bseed is city department (https://detroitmi.gov/departments/buildings-safety-engineering-and-environmental-department)

In [28]:
rental_gdf = gpd.read_file('../data/Rental_Statuses/Rental_Statuses.shp')

In [29]:
len(rental_gdf[rental_gdf['geometry'].isna()])/len(rental_gdf)

0.00803870216790339

In [30]:
len(rental_gdf[rental_gdf['geometry'].isna()])

221

In [39]:
rental_gdf['date_statu'] = pd.to_datetime(rental_gdf['date_statu'])

In [40]:
rental_gdf['date_statu'].max()

Timestamp('2022-05-15 00:00:00')

In [31]:
len(rental_gdf)

27492

In [32]:
rental_gdf[(rental_gdf['street_num']=='120') & (rental_gdf['street_nam']=='SEWARD')]

record_id street_num street_dir street_nam street_typ  date_statu  \
8322   REG2020-08775        120       None     SEWARD       None  2020-12-21   
25324  RNT2020-00032        120       None     SEWARD       None  2020-01-03   

         zip                            record_typ            owner_name  \
8322   48202                   Rental Registration     THOMPSON, BARRY E   
25324  48202  Rental Property Initial Registration  ELIZABETH TINTINALLI   

                owner_addr          owner_city owner_stat owner_zip  \
8322   120 SEWARD UNIT 401             DETROIT         MI     48202   
25324  863 BARRINGTON ROAD  GROSSE POINTE PARK         MI     48230   

                     task  status  parcel_id        lon        lat  ObjectId  \
8322   Issue Registration  Issued  02001199. -83.079031  42.374195      8323   
25324  Issue Registration  Issued  02001199. -83.079031  42.374195     25325   

                         geometry  
8322   POINT (-83.07903 42.37420)  
25324  POINT (-83.07903 42.37420)

- for 18224 hartwell, could be a typo (google maps shows a point right at the lodge and hartwell, without an address there) - we _could_ look donna coulter up in the assessor's db (parcel dataset on open data portal) to see if the address is correct there

In [33]:
rental_gdf[rental_gdf['geometry'].isna()].sample(5)

record_id street_num street_dir                street_nam  \
12290   REG2021-04347       1533       None                    LARNED   
5678    REG2020-06129       4018       None                      29th   
14427   REG2021-06367     5010.0       None                      OPAL   
27482  RNTR2021-00099       5061       None  -67 E OUTER DRIVE BLDG F   
25180   RNT2019-08295      23401       WEST                Eight MILE   

      street_typ  date_statu      zip                            record_typ  \
12290       None  2021-05-21    48207                   Rental Registration   
5678        None  2020-09-08    48221                   Rental Registration   
14427       None  2021-08-05  48236.0                   Rental Registration   
27482       None  2021-05-05    48234  Rental Property Registration Renewal   
25180       None  2019-09-16    48219  Rental Property Initial Registration   

                         owner_name                     owner_addr  \
12290             ORLEANS OWNER LLC    2550 TELEGRAPH RD SUITE 200   
5678         Newberry Homes LDHA-LP  16250 Northland Dr. Suite 301   
14427                 NUNZIO, RUISE                 5010 CANYON ST   
27482  INTERNATIONAL APARTMENTS LLC             32500 W EIGHT MILE   
25180     BONNIEVIEW APARTMENTS LLC   15777 W 10 MILE RD suite 101   

             owner_city owner_stat  owner_zip                task  status  \
12290  BLOOMFIELD HILLS         MI      48302  Issue Registration  Issued   
5678         Southfield         MI      48075  Issue Registration  Issued   
14427     GROSSE POINTE         MI  48236 221  Issue Registration  Issued   
27482        FARMINGTON         MI      48336  Issue Registration  Issued   
25180        SOUTHFIELD         MI      48076  Issue Registration  Issued   

      parcel_id  lon  lat  ObjectId geometry  
12290      None  NaN  NaN     12291     None  
5678       None  NaN  NaN      5679     None  
14427      None  NaN  NaN     14428     None  
27482      None  NaN  NaN     27483     None  
25180      None  NaN  NaN     25181     None

### example for RNTR2021-00099 / 5061 E outer drive bldg F
- 17016320.003 is parcel ID
- dashes (like -67 E outer drive) trips up the geocoder
- we manually looked this up on https://cityofdetroit.github.io/base-unit-tools

### example for REG2021-06367 / 5010 OPAL
- 21078528. is parcel ID
- looking up in base units explorer, it has it as 5010 canyon - bseed thought this was opal (on the corner of opal and canyon) - could be that the assessor thought it was on opal
- we can report this to the assessor's to fix this

## Parcel Data
- https://data.detroitmi.gov/datasets/parcels-2/explore?location=42.352680%2C-83.099134%2C10.81
- fuzzy match on Donna Coulter

In [34]:
parcel_gdf = gpd.read_file('../data/Parcels/Parcels.shp')

In [35]:
parcel_gdf.head()

OBJECTID  object_id parcel_num ward        address council_di zip_code  \
0         1       3084  16014523.   16  4700 CAMPBELL          6    48209   
1         2       3085  16013773.   16   4909 35TH ST          6    48210   
2         3       3086  16014658.   16  4627 CAMPBELL          6    48209   
3         4       3087  16013698.   16   4962 35TH ST          6    48210   
4         5       3088  18002630.   18   6325 PELOUZE          6    48210   

                      taxpayer_1 taxpayer_2          taxpayer_s  ...  \
0                SPEARS, KRYSTAL       None       4700 CAMPBELL  ...   
1           POBEREZHNYY, NIKOLAY       None        4909 35TH ST  ...   
2  URIBE-JUAREZ, MARIA GUADALUPE       None  4627 N CAMPBELL ST  ...   
3             PEREZ, MATEO GOMEZ       None    4962 35TH STREET  ...   
4             MCDERMOTT, LANCE S       None     6681 MERCIER ST  ...   

  assessed_v taxable_va landmap related zoning subdivisio  \
0    10700.0     7777.0     105    None     R2       None   
1    12700.0     7845.0     105    None     R2       None   
2    12900.0     9918.0     105    None     R2       None   
3     8200.0     5672.0     105    None     R2       None   
4      300.0      300.0     105    None     R2       None   

                                          legal_desc SHAPE_Leng    SHAPE_Area  \
0  Property exempt from Ad Valorem taxes and asse...   0.000925  3.259140e-08   
1  W 35TH ST 23 BLK 15-FYFE BARBOUR & WARRENS SUB...   0.000880  3.036103e-08   
2  W--NO CAMPBELL 112 SUB OF PT OF P C 171 L12 P2...   0.001142  4.177216e-08   
3  E 35TH ST 11 BLK 16-FYFE BARBOUR & WARRENS SUB...   0.000880  3.044350e-08   
4  N PELOUZE 116 WESSON & INGERSOLLS SUB L5 P47 P...   0.000905  3.553440e-08   

                                            geometry  
0  POLYGON ((-83.11830 42.33727, -83.11830 42.337...  
1  POLYGON ((-83.11828 42.33796, -83.11796 42.338...  
2  POLYGON ((-83.11828 42.33610, -83.11784 42.336...  
3  POLYGON ((-83.11827 42.33881, -83.11795 42.338...  
4  POLYGON ((-83.12477 42.33497, -83.12467 42.335...  

[5 rows x 41 columns]

In [36]:
parcel_gdf_nona = parcel_gdf[~parcel_gdf['taxpayer_1'].isna()]

In [37]:
parcel_gdf_nona[(parcel_gdf_nona['taxpayer_1'].str.contains('COULTER')) & parcel_gdf_nona['taxpayer_1'].str.contains('DONNA')]

OBJECTID  object_id parcel_num ward        address council_di zip_code  \
33379     33380      37280  16019088.   16  18224 STOEPEL          2    48221   

             taxpayer_1 taxpayer_2        taxpayer_s  ... assessed_v  \
33379  COULTER, DONNA M       None  7390 OLD MILL RD  ...    26200.0   

      taxable_va landmap related zoning subdivisio  \
33379    16094.0     169    None     R1       None   

                                              legal_desc SHAPE_Leng  \
33379  E STOEPEL 371 CANTERBURY GARDEN NO 1 L37 P66 P...   0.000964   

         SHAPE_Area                                           geometry  
33379  3.711491e-08  POLYGON ((-83.14213 42.42488, -83.14175 42.424...  

[1 rows x 41 columns]

- alright, 18224 HARTWELL should've been 18224 STOEPEL - alternate goal is to document any mistakes I find in the geocoding / data entry, and give that to the city. Needs to be added to the db

## Geocoding
- first 2.5k free...?
https://developers.google.com/maps/documentation/geocoding/#Limits